### PR TITLE
Favorites feature

### PR DESCRIPTION
**Pull request by brian-greeson**
- Added favorites controller
 - Added favorite model
 - Added favorite Index view functionality
 - Added favorite counter in nav bar that links to favorite index view
 - Added functionality to prevent adding a favorite pet multiple times: remove favorite link appears instead
 - Added functionality to remove a single or all of favorites
 - Added default favorite index page when there are no favorites
 - [x] Passes all tests?

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -213,4 +211,4 @@ DEPENDENCIES
   web-console (>= 3.3.0)
 
 BUNDLED WITH
-   2.0.1
+   2.1.4
```
Comment (SMJ289) [inline Gemfile.lock]: It is interesting that the Gem file updated. Hopefully it doesn't cause any problems!

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,3 +1,9 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+
+  helper_method :favorite
+
+  def favorite
+    @favorite ||= Favorite.new(session[:favorite])
+  end
 end
```
Comment (SMJ289) [inline app/controllers/application_controller.rb]: A favorites list is a great use case for a session!

File: app/controllers/favorite_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    @pets = []
+    if !favorite.contents.empty?
+      @heading = "My Favorites ❤️"
+      Pet.find_each do |pet|
+        pet_id_string = pet.id.to_s
+        @pets << pet if favorite.contents.include?(pet_id_string)
+      end
+    end
+  end
+
```
Comment (SMJ289) [inline app/controllers/favorite_controller.rb]: Maybe this login in the index method could be refactored and moved to the favorite model?

File: app/controllers/favorite_controller.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  def update
+    pet = Pet.find(params[:pet_id])
+    favorite.add_pet(pet.id)
+    session[:favorite] = favorite.contents
+    flash[:ding_dong] = "#{pet.name} has been added to your favorites list ❤️"
+    redirect_to "/pets/#{params[:pet_id]}"
+  end
+
+  def index
+    @heading = "You have no favorite pets 💔"
```
Comment (SMJ289) [inline app/controllers/favorite_controller.rb]: ❤️

File: app/controllers/pets_controller.rb [hunk 5]
```diff
@@ -37,7 +38,7 @@ def update
       approximate_age: pet_params[:approximate_age],
       sex: pet_params[:sex],
       })
-      
+
     redirect_to "/pets/#{pet.id}"
```
Comment (SMJ289) [inline app/controllers/pets_controller.rb]: We should remove this extra line here.

File: app/models/favorite.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    if contents.include?(pet_id.to_s)
+      favorite_link[:favorite_link_text] = "Remove Favorite"
+      favorite_link[:method] = :delete
+    else
+      favorite_link[:favorite_link_text] = "Add Favorite"
+      favorite_link[:method] = :patch
+    end
+    favorite_link
+  end
+
```
Comment (SMJ289) [inline app/models/favorite.rb]: I like that this logic lives in the model!

File: app/views/favorite/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+<% @pets.each do |pet| %>
+<article id="pet-<%= pet.id %>-details">
+  <img><%= image_tag(pet.image, width: "200") %></img>
+  <h2>Name: <%= link_to "#{pet.name}", "/pets/#{pet.id}"%></h2>
+  <ul>
+    <li><%= link_to 'Remove', "/favorites/#{pet.id}", method: :delete %>
+  </ul>
+  <br>
+</article>
+<% end %>
```
Comment (SMJ289) [inline app/views/favorite/index.html.erb]: The use of id's here is nice. We should add the same functionality in the shelters and pets index.

File: app/views/pets/show.html.erb [hunk 8]
```diff
@@ -5,8 +5,18 @@
...
 <p>Adoption Status: <%= @pet.adoption_status %></p>
-
+<!-- if favorites.include?(pet_id)
+  @favorite_link_text = "Remove Favorite"
+  @favorite_link_method = :delete
+else
+  @favorite_link_text = "Add Favorite"
+  @favorite_link_method = :patch
+end -->
 <ul>
```
Comment (SMJ289) [inline app/views/pets/show.html.erb]: What is this commented out section? Do we need to delete this or uncomment this out and re-run the test?

File: config/routes.rb [hunk 9]
```diff
@@ -10,8 +10,8 @@
   post '/shelters', to: 'shelters#create'
   delete '/shelters/:id', to: 'shelters#destroy'
 
-  get '/pets', to: 'pets#index'
-  get '/pets/:pet_id', to: 'pets#show'
+  get '/pets', to: 'pets#index', as: 'pets_index'
+  get '/pets/:pet_id', to: 'pets#show', as: 'pets_show'
```
Comment (SMJ289) [inline config/routes.rb]: We can try to incorporate these link shortcuts during the refactor.

File: app/controllers/favorite_controller.rb (hunk 3 above)
Comment (brian-greeson) [reply app/controllers/favorite_controller.rb]: I thing thats a great idea. We should open a new branch for that fix

File: app/controllers/pets_controller.rb (hunk 5 above)
Comment (brian-greeson) [reply app/controllers/pets_controller.rb]: Okay

File: spec/features/favorite/add_pet_spec.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      within(".flash-message") do
+        expect(page).to have_content("#{pet_1.name} has been added to your favorites list ❤️")
+      end
+      expect(page).to have_link("Favorites: 1")
+  end
+
+  it "I can  " do
+
+  end
+end
```
Comment (SMJ289) [inline spec/features/favorite/add_pet_spec.rb]: Ummm what is this doing here?

File: spec/features/favorite/index_spec.rb [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+    within "#pet-#{pet_2.id}-details" do
+      expect(page).to have_link("Remove")
+      click_link "Remove"
+    end
+
+    expect(page).to_not have_content(pet_2.name)
+  end
+
+  it "I can visit the favorite index page and see text stating that i have no favorite pets" do
```
Comment (SMJ289) [inline spec/features/favorite/index_spec.rb]: The i should be capitalized in this it...do.

File: app/controllers/favorite_controller.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    session[:favorite] = favorite.contents
+    flash[:ding_dong] = "Pet Removed from Favorites! 💔"
+    redirect_back fallback_location: root_path
+  end
+
+  def destroy_all
+    session[:favorite].clear
+    redirect_to "/favorites"
+  end
+end
```
Comment (SMJ289) [inline app/controllers/favorite_controller.rb]: Even though this is a one-line method should the logic live in the model and simply be called here in the controller?

Comment (brian-greeson) [reply app/controllers/favorite_controller.rb]: I don't remember from the lessons where this is supposed to live. We should ask about this during our next check in

File: app/views/pets/show.html.erb (hunk 8 above)
Comment (brian-greeson) [reply app/views/pets/show.html.erb]: Good catch. That was from us dream driving together.

File: spec/features/favorite/add_pet_spec.rb (hunk 10 above)
Comment (brian-greeson) [reply spec/features/favorite/add_pet_spec.rb]: “¯\_(ツ)_/¯“

**Pull request merged**